File: org.wolfgang/org.wolfgang.contrail-core/src/main/java/org/wolfgang/contrail/component/pipeline/compose/CompositionComponent.java
/*
 * Copyright (C)2012 D. Plaindoux.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation; either version 2, or (at your option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 */

package org.wolfgang.contrail.component.pipeline.compose;

import org.wolfgang.contrail.component.Component;
import org.wolfgang.contrail.component.ComponentConnectionRejectedException;
import org.wolfgang.contrail.component.DestinationComponent;
import org.wolfgang.contrail.component.SourceComponent;
import org.wolfgang.contrail.component.core.AbstractComponent;
import org.wolfgang.contrail.handler.DataHandlerCloseException;
import org.wolfgang.contrail.link.ComponentLinkManager;

/**
 * <code>ComposedPipelineComponent</code>
 * 
 * @author Didier Plaindoux
 * @version 1.0
 */
public class CompositionComponent extends AbstractComponent implements Component {

	private final SourceComponent<?, ?> initialComponent;
	private final DestinationComponent<?, ?> terminalComponent;

	/**
	 * Constructor
	 * 
	 * @throws ComponentConnectionRejectedException
	 */
	public CompositionComponent(ComponentLinkManager linkManager, Component... components) throws ComponentConnectionRejectedException {
		super();

		assert components.length > 1;

		initialComponent = (SourceComponent<?, ?>) components[0];

		for (int i = 1; i < components.length; i++) {
			linkManager.connect(components[i - 1], components[i]);
		}

		terminalComponent = (DestinationComponent<?, ?>) components[components.length - 1];
	}

	@Override
	public void closeUpStream() throws DataHandlerCloseException {
		this.initialComponent.closeUpStream();
	}

	@Override
	public void closeDownStream() throws DataHandlerCloseException {
		this.terminalComponent.closeDownStream();
	}
}

File: org.wolfgang/org.wolfgang.contrail-js/src/main/js/Contrail/component/composition/CompositionComponent.js
/*
 * Copyright (C)2012 D. Plaindoux.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation; either version 2, or (at your option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 */

/*global define:true, require, module*/

if (typeof define !== "function") {
    var define = require("amdefine")(module);
}

define(["require", "Core/object/jObj" ],
    function (require, jObj) {
        "use strict";

        function CompositionComponent(components) {
            jObj.bless(this, require("Contrail/component/jComponent").core.component());
            this.components = components;
        }

        /**
         * Constructor
         *
         * @type {*|void}
         */
        CompositionComponent.init = jObj.constructor([ jObj.types.ArrayOf(jObj.types.Named("Component")) ],
            function (components) {
                return new CompositionComponent(components);
            });

        CompositionComponent.prototype.getSource = jObj.method([], jObj.types.Named("SourceComponent"),
            function () {
                return this.components[this.components.length - 1].getSource();
            });

        CompositionComponent.prototype.closeDownStream = jObj.procedure([],
            function () {
                this.getSource().closeDownStream();
            });

        CompositionComponent.prototype.getDownStreamDataFlow = jObj.method([], "DataFlow",
            function () {
                return this.getSource().getDownStreamDataFlow();
            });

        CompositionComponent.prototype.getDestination = jObj.method([], jObj.types.Named("DestinationComponent"),
            function () {
                return this.components[0].getDestination();
            });

        CompositionComponent.prototype.closeUpStream = jObj.procedure([],
            function () {
                this.getDestination().closeUpStream();
            });

        CompositionComponent.prototype.getUpStreamDataFlow = jObj.method([], "DataFlow",
            function () {
                return this.getDestination().getUpStreamDataFlow();
            });

        return CompositionComponent.init;
    });
